docs(guides): update next.js tech guide to the relevant status (#830)
feature-sliced/documentation: i18n/en/docusaurus-plugin-content-docs/current/guides/tech/with-nextjs.mdx

This routing mechanism **does not align** with the FSD concept, as it's not possible to maintain a flat slice structure with such a routing mechanism.

The approach is to move the Next.js `app` folder to the project root and import FSD pages into the Next.js `app` folder.
This preserves the FSD project structure inside the `src` folder, where layer folders should be placed.

You should also add a `pages` folder to the project root because App Router is compatible with Pages Router.
The best way is to put `README.md` file inside describing the need to save this folder.

```sh
├── app                              # App folder (Next.js)
    ├── api
    │   └── get-example
    │       └── route.ts
│   └── example
│       └── page.tsx
├── pages                            # Legacy pages folder (Next.js)
│   └── README.md                    # To maintain compatibility with Pages Router
└── src
    ├── app
    ├── pages
    │   └── example
    │       ├── index.ts
    │       ├── ui
    │       │   └── example.tsx
    │       └── routes
    │           └── get-example.ts   # Route handler
    ├── widgets
    ├── features
    ├── entities
    └── shared
```

> ### Example of re-exporting a page from `src/pages/example` to `app/example/page.tsx`

Reviewer: issue: тут тоже раздел получается немного неописательным, и на мой взгляд, незаслуженно отделенным от прошлого раздела.

suggestion: мб не делать эти примеры заголовками, а просто текстом обычным оставить?
```suggestion
Example of re-exporting a page from `src/pages/example` to `app/example/page.tsx`:
```

```tsx title="app/example/page.tsx"
export { Example as default, metadata } from '@/pages/example';
```

### Middleware
If you use middleware in your project, it must be located in the project root alongside the `app` and `pages` folders.

### Instrumentation
The `instrumentation.ts|js` file allows you to monitor the performance and behavior of your application. If you use it, it must be located in the project root, similar to `middleware.ts|js`.

### Route Handlers
It is suggested to use a new `routes` segment in the `app` or `pages` layer to work with Route Handlers.

```tsx title="app/routes/get-example/route.ts"
export { getExample as GET } from '@/pages/example';
```

### Additional recommendations

> - Use the `api` segment in the `shared` layer to describe database queries and their further use in higher layers.

Reviewer: issue: Пока писал прошлый коммент, понял, что мне все-таки не очень нравится идея описывать какие-то переиспользуемые запросы к БД в сегменте `api`, это все-таки совсем другое. 

suggestion: Как насчет порекомендовать тут тоже использовать сегмент `db`? Будет тогда следующая параллель:
- в корне сегмента shared/api обычно описывается какой-то API-клиент, который используется другими, более конкретными API-функциями в shared/api/<endpoint> (сгруппированных по эндпоинту)
- в корне сегмента shared/db будет описываться клиент БД, будь то дриззл или призма или еще че, и он затем будет использоваться более конкретными функциями-запросами, сгруппированными по основной таблице или бизнес-сущности

Reviewer: Тогда и следующую строчку тоже надо будет обновить

- Caching and revalidating queries logic is better kept in the same place as the queries themselves, in a basic scenario this is the `api` segment in the `shared` layer.

## Legacy approach with Pages Router

> Page routers should be placed in the `pages` folder in the root of the project, similar to `app` for the App Router.

Reviewer: suggestion (non-blocking):
```suggestion
When using the Pages Router, routes should be placed in the `pages` folder in the root of the project, similar to `app` for the App Router.
```

The structure inside `src` where the layer folders are located remains unchanged.

```sh
├── pages                        # Pages folder (Next.js)
│   ├── _app.tsx
│   ├── api
│   │   └── example.ts           # API Route re-export
│   └── example
│       └── index.tsx
└── src
    ├── app
    ├── pages
    │   └── example
    │       ├── index.ts
    │       ├── ui
    │       │   └── example.tsx
    │       └── routes
    │           ├── config.ts    # API Route config
    │           └── handler.ts   # API Route
    ├── widgets
    ├── features
    ├── entities
    └── shared
```

> ### Functionality re-export from the public API `src/app/index.ts` to `pages/_app.tsx`

Reviewer: issue: это какая-то не очень понятная секция, без текста, только кодовый блок. Помимо этого, мне кажется, Steiger будет ругаться на индекс-файл в app, но я не уверен

suggestion: хочется найти какое-то более стабильное место кастомному компоненту App. При этом, не хочется выдумывать для него другие имена, которые не будут знакомы Next-разработчикам. Может, забить и просто предложить людям создать в слое `app` сегмент  `_app`? Повторение, конечно, но зато сразу понятно всем, кто знают Pages Router. Альтернативно можно предложить `custom-app`

И давай заголовок этой секции сделаем вот такой:

```suggestion
### Custom `_app` component
```

И чуть текста добавим, типа "You can place your Custom App component in `src/app/_app` or `src/app/custom-app`, и затем ниже привести опять же два код-блока — один с содержимым кастомного аппа, другой с реэкспортом в нужное место

│   ├── api
│   │   └── get-example
│   │       └── route.ts
│   └── example
│       └── page.tsx
├── pages                            # Legacy pages folder (Next.js)
│   └── README.md                    # To maintain compatibility with Pages Router
└── src
    ├── app
    ├── pages
    │   └── example
    │       ├── index.ts
    │       ├── ui
    │       │   └── example.tsx
    │       └── routes
    │           └── get-example.ts   # Route handler
    ├── widgets
    ├── features
    ├── entities
    └── shared
```

### Example of re-exporting a page from `src/pages/example` to `app/example/page.tsx`

```tsx title="app/example/page.tsx"
export { Example as default, metadata } from '@/pages/example';
```

### Middleware
If you use middleware in your project, it must be located in the project root alongside the `app` and `pages` folders.

### Instrumentation
The `instrumentation.ts|js` file allows you to monitor the performance and behavior of your application. If you use it, it must be located in the project root, similar to `middleware.ts|js`.

### Route Handlers
It is suggested to use a new `api-routes` segment in the `app` layer to work with Route Handlers.

> However, you should avoid overusing them due to code maintenance complexity and violation of separation of concerns principles.

Reviewer: could you also update that in the Russian version please?

---
sidebar_position: 10
---
# Usage with Next.js

It's possible to implement FSD in a Next.js project, but conflicts arise due to differences between Next.js project structure requirements and FSD principles.

## Conflict between FSD and Next.js in the `app` layer

> Next.js suggests using the `app` folder to define application routes. Next.js expects files in the `app` folder to correspond to URLs.

Reviewer: I see that it's not you who introduced these issues, but thought that now's a good time to fix them anyway

This routing mechanism **does not align** with the FSD concept, as it's not possible to maintain a flat slice structure with such a routing mechanism.

The approach is to move the Next.js `app` folder to the project root and import FSD pages into the Next.js `app` folder.
This preserves the FSD project structure inside the `src` folder.

You should also add a `pages` folder to the project root because App Router is compatible with Pages Router.

> The best way is to put an empty `.gitkeep` or `README.md` file inside describing the need to save this folder.

Reviewer: suggestion (non-blocking): I would focus the recommendation to the README file instead of `.gitkeep` as it's not obvious why this empty folder is needed, so project docs could really come in handy later down the line

```sh
├── app              # App folder (Next.js)
├── pages            # Legacy pages folder (Next.js)
│   └── .gitkeep     # To maintain compatibility with Pages Router
└── src
    ├── app
    ├── pages
    ├── widgets
    ├── features
    ├── entities
    └── shared
```

### Example of re-exporting a page from `src/pages/example` to `app/example/page.tsx`

```tsx title="app/example/page.tsx"
export { Example as default, metadata } from '@/pages/example';
```

### Middleware
If you use middleware in your project, it must be located in the project root alongside the `app` and `pages` folders.

### Instrumentation
The `instrumentation.ts|js` file allows you to monitor the performance and behavior of your application. If you use it, it must be located in the project root, similar to `middleware.ts|js`.

### Route Handlers

> It is suggested to use a new `routes` segment in the `app` or `pages` layer to work with Route Handlers.

Reviewer: issue: this part isn't clear to me. You mention that it's recommended to use the `routes` segment, but the code block below doesn't demonstrate it.

suggestion: let's also add a code block where `getExample` is defined so that it's clear how and where to define route handlers

But also, should this really be in the FSD structure? I feel like mixing in unrelated backend code into the frontend code folder might make it hard to find. Let's discuss and refine this point a bit, perhaps also get people from the chat to give their opinion

Reviewer: В целом, я согласен с тобой, что есть польза в том, чтоб описать, куда мы предлагаем класть простенькие API-роуты, раз уж мы всю структуру перерыли. Давай может тогда вот так поступим:
- уберем из рекомендаций предложение хранить их в pages — они не имеют отношения к страницам, пусть уж будут в `app` только
- сделаем два код-блока в этой секции, первый в `src/app/api-routes` (мне больше нравится api-routes, потому что тогда не путаем с роутами, по которым находятся страницы), и в нем прямо полноценное описание функции-хендлера `getExampleData`. Внутри этой функции можем даже заюзать `db` из `shared/db`, чтоб продемонстрировать реюз коннекшна между роут-хендлерами и серверными компонентами
- второй код-блок в `app/api/example`, и он реэкспортирует `getExampleData as GET`

```tsx title="app/api/get-example/route.ts"
export { getExample as GET } from '@/pages/example';
```

### Additional recommendations

> - Use the `queries` segment in the `shared` layer to describe database queries and their further use in higher layers.

Reviewer: issue: I think that `queries` here describes what the code is, not what the code is for (see https://github.com/feature-sliced/steiger/blob/master/packages/steiger-plugin-fsd/src/segments-by-purpose/README.md)

suggestion: I prefer to keep the logic of server actions and data loaders in the `api` segment of pages that need it, but let's discuss this in more detail too

- Caching and revalidating queries logic is better kept in the same place as the queries themselves, in a basic scenario this is the `api` segment in the `shared` layer.

## Legacy approach with Pages Router

In Next.js below version 13, there is no explicit `app` folder. Next.js provides a `_app.tsx` file instead, which serves as a wrapper component for all project pages.

### Importing functionality into `pages/_app.tsx`

> To solve the problem of missing the `app` folder in the Next.js structure, you can create an `App` component inside the `app` layer and import it into `pages/_app.tsx` so that Next.js can work with it.

Reviewer: issue: to be honest, this section doesn't really make sense, I don't remember why it was added

suggestion: maybe we should rewrite it with proper guidelines for the Pages router?

Author: Yep, I will rewrite it

The approach is to move the NextJS `app` folder to the root folder of the project and import the FSD pages into the NextJS `app` folder. This saves
the FSD project structure inside the `src` folder. You should still also add the `pages` folder to the root, because the App router is compatible with the Pages router.

```
├── app                # NextJS app folder
├── pages              # Stub NextJS pages folder
│   ├── README.md      # Description of why this folder exists
├── src
│   ├── app            # FSD app folder
│   ├── entities
│   ├── features
│   ├── pages          # FSD pages folder
│   ├── shared
│   ├── widgets
```

[![Open in StackBlitz](https://developer.stackblitz.com/img/open_in_stackblitz.svg)][ext-app-router-stackblitz]

## Middleware

If you are using middleware in a project, it also needs to be moved from `src` to the root of the project. Otherwise, NextJS simply won't see it — middleware must be located strictly at the root next to `app` or `pages`.

## See also {#see-also}

- [(Thread) About the pages directory in NextJS](https://t.me/feature_sliced/3623)
## See also

[project-knowledge]: /docs/about/understanding/knowledge-types

> [ext-app-router-stackblitz]: https://stackblitz.com/edit/stackblitz-starters-aiez55?file=README.md

Reviewer: suggestion: let's not remove this example, I think it's useful for people to see a functioning app that implements this recommendation. If there's something you'd like to change about it — let me know, I have write access to it
